README.md
X Knife Spork
=============
X Knife Spork is a tool for manipulating infrastructure managed by chef.

This is a modified version of the original [knife-spork project](https://github.com/jonlives/knife-spork) by [@jonlives](https://github.com/jonlives) at Etsy.

Installation
------------
```bundle install```

OR

```bash
git clone git@github.com:AppLovin/knife-spork.git
gem build knife-spork.gemspec
gem install knife-spork-2.9.0.gem
```

Usage
------------
### Change attributes in multiple environments 

```
knife spork environment attribute set ENVIRONMENT ATTRIBUTE VALUE
knife spork environment attribute unset ENVIRONMENT ATTRIBUTE
```

###### Example 1: Pass a list of environments

```
$ knife spork environment attribute set Environment1,Environment2 hello world
Modifying Environment1
Done modifying Environment1 at 2015-08-06 23:55:06 -0700
Modifying Environment2
Done modifying Environment2 at 2015-08-06 23:55:07 -0700
```

This will add the key ```hello``` and set it with the value of ```world``` in the override_attributes section of the Environment1 and Environment2: 

```json
{
  "override_attributes": {
    "hello": "world"
  }
}
```

###### Example 2: Pass the name of a group of environments as specified in the configuration file
```
$ knife spork environment attribute set test hello world
Modifying Environment1
Done modifying Environment1 at 2015-08-06 23:55:06 -0700
Modifying Environment2
Done modifying Environment2 at 2015-08-06 23:55:07 -0700
```

### Unset environment attributes

This is the opposite of ```environment attribute set``` subcommand. It removes an attribute from the environment

```bash
$ knife spork environment attribute unset test hello
Modifying TestEnvironment4
Done modifying TestEnvironment4 at 2015-08-16 13:36:48 -0700
Modifying TestEnvironment5
Done modifying TestEnvironment5 at 2015-08-16 13:36:48 -0700
```

Spork Info
-----------
This function is designed to help you see which plugins you currently have loaded, and the current config Hash which knife spork is using.

#### Usage
```bash
knife spork info
```

#### Example

```text
$ knife spork info
Config Hash:
{"plugins"=>{"git"=>{"enabled"=>true}, "irccat"=>{"server"=>"irccat.mydomain.com", "port"=>12345, "gist"=>"usr/bin/gist", "channel"=>["#chef-announce"]}, "graphite"=>{"server"=>"graphite.mydomain.com", "port"=>2003}, "slack"=>{"api_token"=>"abc123", "channel"=>"random", "teamname"=>"myteam", "icon_url"=>"http://example.com/image.jpg"}, "eventinator"=>{"url"=>"http://eventinator.mydomain.com/events/oneshot"}}, "default_environments"=>["development", "production"], "version_change_threshold"=>2, "pplugins"=>{"foodcritic"=>{"fail_tags"=>["style,correctness,test"], "tags"=>["~portability"], "include_rules"=>["config/rules.rb"]}}}

Plugins:
KnifeSpork::Plugins::Campfire: disabled
KnifeSpork::Plugins::Eventinator: enabled
KnifeSpork::Plugins::Foodcritic: disabled
KnifeSpork::Plugins::Git: enabled
KnifeSpork::Plugins::Graphite: enabled
KnifeSpork::Plugins::HipChat: disabled
KnifeSpork::Plugins::Slack: enabled
KnifeSpork::Plugins::Irccat: enabled
```

Spork Check
-----------
This function is designed to help you avoid trampling on other people's cookbook versions, and to make sure that when you come to version your own work it's easy to see what version numbers have already been used and if the one you're using will overwrite anything.

#### Usage
```bash
knife spork check COOKBOOK [--all]
```

By default, spork check only shows the 5 most recent remote cookbook versions. Add the --all option if you want to see everything.

#### Example (Checking an Unfrozen Cookbook with version clash)

```text
$ knife spork check apache2
Checking versions for cookbook apache2...

Local Version:
  1.1.49

Remote Versions: (* indicates frozen)
 *2.0.2
 *2.0.1
  1.1.49
 *1.1.14
 *1.1.13

ERROR: The version 1.1.49 exists on the server and is not frozen. Uploading will overwrite!
```

#### Example (Checking a Frozen Cookbook with version clash)

```text
$ knife spork check apache2
Checking versions for cookbook apache2...

Local Version:
  2.0.2

Remote Versions: (* indicates frozen)
 *2.0.2
 *2.0.1
  1.1.49
 *1.1.14
 *1.1.13

WARNING: Your local version (2.0.2) is frozen on the remote server. You'll need to bump before you can upload.
````

#### Example (No version clashes)

```text
$ knife spork check apache2
Checking versions for cookbook apache2...

Local Version:
  2.0.3

Remote Versions: (* indicates frozen)
 *2.0.2
 *2.0.1
  1.1.49
 *1.1.14
 *1.1.13

Everything looks good!
```

Spork Bump
----------
This function lets you easily version your cookbooks without having to manually edit the cookbook's `metadata.rb` file. You can either specify the version level you'd like to bump (`major`, `minor`, or `patch`), or you can manually specify a version number. This might be used if, for example, you want to jump several version numbers in one go and don't want to have to run knife bump once for each number. If no bump level is specified, a patch level bump will be performed.

#### Usage
```bash
knife spork bump COOKBOOK [major | minor | patch | manual x.x.x]
````

#### Example (No patch level specified - defaulting to patch)
```text
$ knife spork bump apache2
Successfully bumped apache2 to v2.0.4!
```

#### Example (Bumping patch level)
```text
$ knife spork bump apache2 patch
Successfully bumped apache2 to v2.0.4!
```

#### Example (Manually setting version)
```text
$ knife spork bump apache2 manual 1.0.13
Successfully bumped apache2 to v1.0.13!
```

Spork Upload
------------
This function works mostly the same as normal `knife cookbook upload COOKBOOK` except that this automatically freezes cookbooks when you upload them.

#### Usage
```bash
knife spork upload COOKBOOK
```
#### Example
```text
$ knife spork upload apache2
Freezing apache2 at 1.0.13...
Successfully uploaded apache2@1.0.13!
```

Spork Promote
-------------
This function lets you easily set a version constraint in an environment or group of environments for a particular cookbook. By default it will set the version constraint to whatever the local version of the specified cookbook is. Optionally, you can include a `--version` option which will set the version constraint for the specified cookbook to whatever version number you provide. You might want to use this if, for example, you pushed a version constraint for a cookbook version you don't want your nodes to use anymore, so you want to "roll back" the environment to a previous version. You can also specify the `--remote` option if you'd like to automatically upload your changed local environment file to the server. To promote to multiple environments, simply pass a list of them delimited with a comma.  

If you don't specify an environment or environment group, the default_environments config directive will be used if set.

#### Usage

```bash
knife spork promote [ENVIRONMENT OR ENVIRONMENT GROUP NAME] COOKBOOK [--version, --remote]
```

#### Example (Using local cookbook version number)

```text
$ knife spork promote my_environment apache2 --remote
Adding version constraint apache2 = 1.0.13
Saving changes to my_environment.json
Uploading my_environment to Chef Server
Promotion complete!
```

#### Example (Using manual version)
```text
$ knife spork promote my_environment apache2 -v 2.0.2
Adding version constraint apache2 = 2.0.2
Saving changes to my_environment.json
Promotion complete. Don't forget to upload your changed my_environment to Chef Server
```

Spork Omni
-------------
Omni lets you combine one of the most common combinations of spork commands (bump, upload & promote or promote --remote) - into one handy shortcut.

As omni is designed for use only in those cases where you want to perform all three of bump, upload and promote at the same time it supports a limited subset of the command line options supported by the individual bump, upload and promote commands.

If you run omni with no extra options, it will default to performing a ```patch``` level bump, and promote locally to the environments listed in the ```default_environments``` variable in your spork configuration file.

Alternatively, you can specify any of the following options:

```--cookbook-path PATH:PATH```: A colon-separated path to look for cookbooks in

```--include-dependencies```: Also upload cookbook dependencies during the upload step

```--bump-level [major|minor|patch]```: Version level to bump the cookbook (defaults to patch)

```--environment ENVIRONMENT```: Environment to promote the cookbook to',

```--remote```: Make omni perform a promote --remote instead of a local promote',

#### Usage

```bash
knife spork omni COOKBOOK [--bump-level, --cookbook-path, --include-dependencies, --environment, --remote]
```

#### Example (default options, default_environments set to development and production)

```text
$ knife spork omni apache2
OMNI: Bumping apache2
Successfully bumped apache2 to v0.3.99!

OMNI: Uploading apache2
Freezing apache2 at 0.3.99...
Successfully uploaded apache2@0.3.99!

OMNI: Promoting apache2
Adding version constraint apache2 = 0.3.99
Saving changes to development.json
Promotion complete. Don't forget to upload your changed development.json to Chef Server
Adding version constraint apache2 = 0.3.99
Saving changes to production.json
Promotion complete. Don't forget to upload your changed production.json to Chef Server
```

#### Example (default options, default_environments set to development and production, promote --remote)

```text
$ knife spork omni apache2 --remote
OMNI: Bumping apache2
Successfully bumped apache2 to v0.3.99!

OMNI: Uploading apache2
Freezing apache2 at 0.3.99...
Successfully uploaded apache2@0.3.99!

OMNI: Promoting apache2
Adding version constraint apache2 = 0.3.99
Saving changes to development.json
Uploading development.json to Chef Server
Promotion complete at 2013-08-08 11:43:12 +0100!
Adding version constraint apache2 = 0.3.99
Saving changes to production.json
Uploading production.json to Chef Server
Promotion complete at 2013-08-08 11:43:12 +0100!
```

#### Example (Specifying patch level and environment)
```text
$ knife spork omni apache2 -l minor -e development
OMNI: Bumping apache2
Successfully bumped apache2 to v0.4.0!

OMNI: Uploading apache2
Freezing apache2 at 0.4.0...
Successfully uploaded apache2@0.4.0!

OMNI: Promoting apache2
Adding version constraint apache2 = 0.4.0
Saving changes to development.json
Promotion complete. Don't forget to upload your changed development.json to Chef Server
```

Spork Environment Check
-------------
Environment check provides the ability to validate a local chef environment file is locked to versions that actually exist on the chef server.  Running this check prior to uploading your environment files is recommended as it can prevent your chef server from becoming unresponsive due to a version constraint that will never be valid.

If you run environment check with no options it will collect all invalid cookbook versions before reporting a failure.

Alternatively, you can specify any of the following options:

```-f, --fatal```: Quit on first invalid constraint located

#### Usage

```bash
knife spork environment check ENVIRONMENT (options)
```

#### Example of a passing chef environment check with no options

```text
$ knife spork environment check production

Checking constraints for environment: production
Environment production looks good
```

#### Example of multiple invalid cookbooks with no options

```text
$ knife spork environment check production

Checking constraints for environment: production
ERROR: mysql@10.2.0 does not exist on Chef Server! Upload the cookbook first by running:

	knife spork upload mysql

ERROR: rbenv@10.7.1 does not exist on Chef Server! Upload the cookbook first by running:

	knife spork upload rbenv

FATAL: Environment production has constraints that point to non existent cookbook versions.
```

#### Example of running with the --fatal flag with the same invalid cookbooks from previous example

```text
$ knife spork environment check production

Checking constraints for environment: production
FATAL: mysql@10.2.0 does not exist on Chef Server! Upload the cookbook first by running:

	knife spork upload mysql
```

Spork Node / Role / Databag Commands
-------------

Knife spork contains "wrappers" around several stock knife commands for changing nodes, roles, environments and databags. These commands work exactly as in the original knife, with the addition of being wrapped in Spork's plugin API. This means that for example, when you upload a role using spork you will see an IRC message containing details, and a gist of the role change.

The following  ```data bag``` commands are provided in knife-spork:
```
knife spork data bag create
knife spork data bag delete
knife spork data bag edit
knife spork data bag from file
```

The following  ```node``` commands are provided in knife-spork:
```
knife spork node create
knife spork node delete
knife spork node edit
knife spork node from file
knife spork node run_list add
knife spork node run_list remove
knife spork node run_list set
```

The following  ```role``` commands are provided in knife-spork:
```
knife spork role create
knife spork role delete
knife spork role edit
knife spork role from file
```

The following  ```environment``` commands are provided in knife-spork:
```
knife spork environment create
knife spork environment delete
knife spork environment edit
knife spork environment from file
```

Spork Configuration
-------------------
Out of the box, knife spork will work with no configuration. However, you can optionally enable several features to enhance its functionality.

KnifeArchon will look for a configuration file in the following locations, in ascending order of precedence:

- `config/spork-config.yml`
- `/etc/spork-config.yml`
- `~/.chef/spork-config.yml`

Anything set in the configuration file in your home directory for example, will override options set in your Chef repository or `/etc`.

Below is a sample config file with all supported options and all shipped plugins enabled below, followed by an explanation of each section.

```yaml
default_environments:
  - development
  - production
environment_groups:
  qa_group:
    - quality_assurance
    - staging
  test_group:
    - user_testing
    - acceptance_testing
version_change_threshold: 2
preserve_constraint_operators: true
environment_path: "/home/me/environments"
save_environment_locally_on_create: false
role_path: "/home/me/roles"
custom_plugin_path: "/home/me/spork-plugins"
always_promote_remote: true
skip_berkshelf: false
role_match_file_name: true
json_options:
  indent: "    "
plugins:
  campfire:
    account: myaccount
    token: a1b2c3d4...
  foodcritic:
    tags: ['any']
  hipchat:
    api_token: ABC123
    rooms:
      - General
      - Web Operations
    notify: true
    color: yellow
  jabber:
    username: YOURUSER
    password: YOURPASSWORD
    nickname: Chef Bot
    server_name: your.jabberserver.com
    server_port: 5222
    rooms:
      - engineering@your.conference.com/spork
      - systems@your.conference.com/spork
  git:
    enabled: true
    auto_push: true
    branch: some_branch
  irccat:
    server: irccat.mydomain.com
    port: 12345
    gist: "/usr/bin/gist"
    channel: ["chef-annoucements"]
  graphite:
    server: graphite.mydomain.com
    port: 2003
  influxdb:
    database: deployments
    username: deploy
    password: deploy
    series: deployments
    host: influx.example.com
    port: 8086
  eventinator:
    url: http://eventinator.mydomain.com/events/oneshot
  slack:
    api_token: abc123
    channel: "#sysops"
    teamname: myteam
    username: knife
    icon_url: http://example.com/image.jpg
  rubocop:
    epic_fail: true
    show_name: false
    autocorrect: false
    out_file: <file>
    sev_level: <C|W|E>
    lint: false
```

#### Default Environments
The `default_environments` directive allows you to specify a default list of environments you want to promote changes to. If this option is configured and you *omit* the environment parameter when promoting KnifeArchon will promote to all environments in this list.

#### Environment Groups
The `environment_groups` directive allows you to specify a list of environments referenced by group names that you want to promote changes to.

#### Version Change Threshold
The `version_change_threshold` directive allows you to customise the threshold used by a safety check in spork promote which will prompt for confirmation if you're promoting a cookbook by more than version_change_threshold versions. This defaults to 2 if not set, ie promoting a cookbook from v1.0.1 to v 1.0.2 will not trip this check, wheras promoting from v1.0.1 to v1.0.3 will.

#### Preserve Constraint Operators
The `preserve_constraint_operators` directive causes spork promote to preserve existing version constraint operators in your environment files, only updating the version number. This directive is disabled by default, which causes spork to always use the `=` constraint.

#### Always Promote Remote
The `always_promote_remote` directive allows you to tell spork promote to always act as if the --remote option had been specified. This will also have the same effect on spork omni. This option should only be used if you're sure you want all changes to be uploaded to the server as soon as you run promote.

#### Skip Berkshelf
The `skip_berkshelf` directive is a temporary flag added in [#138](https://github.com/jonlives/knife-spork/issues/138) to allow Berkshelf functionality to be optionally bypassed until Berkshelf 3 support has been added to knife-spork per [#85](https://github.com/jonlives/knife-spork/issues/85). It simply removed the :Berkshelf constant from the namespace used by knife-spork.

#### JSON Options

CHANGELOG.md
## 2.10.3(25th October, 2017)

Bugfixes:

    - Modify attributes in default section instead in `environment attribute set` subcommand

## 2.10.2(25th October, 2017)

Bugfixes:

    - Used different config attribute to get disabled git hooks for auto push


## 2.10.1(10th October, 2017)

Bugfixes:

    - Fixed slack messaging in multichef after_environmentfromfile


## 2.10.0(10th October, 2017)

Features:

    - Add multichef hook for spork role from file

## 2.9.0(9th October, 2017)

Features:

    - Add multichef hook for spork node edit


## 2.8.0(29th September, 2017)

Features:

    - Add multichef hook for spork environment from file

## 2.7.0(27th September, 2017)

Features:

    - Add slack messaging for multichef plugin

## 2.6.0(27th September, 2017)

Features:

    - Initial commit for multichef plugin

## 2.5.2(21st February, 2017)

Refactoring:

    - Reimplemented creating a branch name

## 2.5.1(8th February, 2017)

Features:

    - Add time to branch name in git plugin

## 2.5.0(15th October, 2016)

Fork:

    - First clean release after forking

## 1.5.1(26th February, 2015)

Bugfixes:

    - Fix 'undefined local variable' error in Hipchat plugin (Thanks to @danieleva https://github.com/jonlives/knife-spork/pull/177)


## 1.5.0(30th January, 2015)

Features:

    - InfluxDB Plugin added (Thanks to @vic3lord https://github.com/jonlives/knife-spork/pull/167)
    - Add gist & diff behavior to the Hipchat plugin (Thanks to @danieleva https://github.com/jonlives/knife-spork/pull/153)
    - Optional config flag to enforce that role files share the same name as the role they contain (Thanks to @nishansubedi https://github.com/jonlives/knife-spork/pull/152)
    - Add optional --cookbook-path flag to Spork Promote (Thanks to @bcandrea https://github.com/jonlives/knife-spork/pull/169)
    - Add --cookbook-path option to spork promote command (Thanks to @bcandrea https://github.com/jonlives/knife-spork/pull/169)
    - Add verification check to environment create commadn when environment already exists (https://github.com/jonlives/knife-spork/issues/171)
    - Add save_environment_locally_on_create option to have environment create command save a local copy of the environment file (https://github.com/jonlives/knife-spork/issues/172)
    - Lazily load Berkshelf to improve overall knife performance (Thanks to @lamont-granquist https://github.com/jonlives/knife-spork/pull/158)

Bugfixes:

    - Add cookbook name metadata field to test cookbook (Thanks to @jdmundrawala https://github.com/jonlives/knife-spork/pull/162)
    - Add missing require statements to fix specs running against Chef 12 (Thanks to @jdmunrawala https://github.com/jonlives/knife-spork/pull/163)
    - Fix issues with the slack plugin (Thanks to @therobot https://github.com/jonlives/knife-spork/pull/160)
    - Fix issue with knife spork omni and skip_berkshelf option when Berkshelf not installed (Thanks to @ctrlok https://github.com/jonlives/knife-spork/pull/168)
    - Fix yajl error on missing environment files (Thanks to @oker1 https://github.com/jonlives/knife-spork/pull/165)
    - Fix for changed constant name from Yaj to FFI_Yajl in Chef 12 (Thanks to @jeunito https://github.com/jonlives/knife-spork/pull/154)
    - Fix Spork DataBag Create to handle specifying a data bag item name too (https://github.com/jonlives/knife-spork/issues/156)


## 1.4.2(3rd November, 2014)

Features:

    - Improve error messages when uploading invalid JSON (Thanks to @jeunito https://github.com/jonlives/knife-spork/pull/151)
    - Allow an optional comma delimited set of environments to be passed to spork promote (Thanks to @jeunito https://github.com/jonlives/knife-spork/pull/148)
    - Add support for rubocop >= 0.23.0 in RuboCop plugin (Thanks to @dwradcliffe https://github.com/jonlives/knife-spork/pull/147)
    - Relax foodcritic version requirement from ~> 3.0.0 to >= 3.0.0 (Thanks to @dwradcliffe https://github.com/jonlives/knife-spork/pull/146)
    - Slack plugin now supports adding images in messages (Thanks to @ctrlok https://github.com/jonlives/knife-spork/pull/144)
    - Git plugin can now optionally commit and push on changes (Thanks to @jeunito https://github.com/jonlives/knife-spork/pull/124)
    
Bugfixes:

    - Fix custom plugin loading on Window (Thanks to @carpnick https://github.com/jonlives/knife-spork/pull/150)
    - Fix config loading in CookbookUploader under Chef 12 (Thanks to @jordane https://github.com/jonlives/knife-spork/pull/149)
    - Do not attempt to load Berksfile if not present (Thanks to @redondos https://github.com/jonlives/knife-spork/pull/142)

## 1.4.1 (21st August, 2014)

Bugfixes:

    - Fix bug with preserving constraint operators on new cookbooks (https://github.com/jonlives/knife-spork/issues/139)
    

## 1.4.0 (21st August, 2014)

Bugfixes:

    - Fix bug with preserve_constraint_operators functionality (https://github.com/jonlives/knife-spork/issues/131)
    - Fix incorrect data bag banner message (Thanks to @jperry https://github.com/jonlives/knife-spork/pull/119)
    - Fix error when calculating environment path when "cookbooks" occurs in cookbook_path outside of the last position (https://github.com/jonlives/knife-spork/issues/137)
    - Fix spelling error in README (Thanks to @jrwesolo https://github.com/jonlives/knife-spork/pull/136)

Features:
    
    - Hipchat plugin now supports custom server URL and API version (Thanks to @hrak https://github.com/jonlives/knife-spork/pull/135)
    - Slack plugin (Thanks to @chrisferry https://github.com/jonlives/knife-spork/pull/134)
    - Rubocop plugin (Thanks to @chazzly https://github.com/jonlives/knife-spork/pull/120)
    - Add skip_berkshelf flag to allow Berkshelf operations to be skipped even when Berkshelf is loaded (https://github.com/jonlives/knife-spork/issues/138)
    
    
## 1.3.4 (2nd June, 2014)

Bugfixes:


lib/knife-spork/runner.rb
require 'app_conf'
require 'diffy'
require 'json'

require 'chef/cookbook_loader'
require 'chef/environment'
require 'chef/data_bag_item'
require 'chef/knife/core/object_loader'
require 'knife-spork/plugins'

module KnifeSpork
  module Runner
    module ClassMethods; end

    module InstanceMethods

      def spork_config
        return @spork_config unless @spork_config.nil?

        @spork_config = AppConf.new
        load_paths.each do |load_path|
          if File.exists?(load_path)
            @spork_config.load(load_path)
          end
        end

        @spork_config
      end

      def run_plugins(hook)
        cookbooks = [ @cookbooks || @cookbook ].flatten.compact.collect{|cookbook| cookbook.is_a?(::Chef::CookbookVersion) ? cookbook : load_cookbook(cookbook)}.sort{|a,b| a.name.to_s <=> b.name.to_s}

        # Affects promote only:
        # Set loaded cookbook version if the -v or --version parameter was specified
        # Otherwise the version on disk, often more recent will be used.
        # We know cookbooks will only contain one cookbook in the case of promote.
        cookbooks.map{|c|c.version = config[:version]} if config[:version]

        environments = [ @environments || @environment ].flatten.compact.collect{|environment| environment.is_a?(::Chef::Environment) ? environment : load_environment_from_file(environment)}.sort{|a,b| a.name.to_s <=> b.name.to_s}
        environment_diffs = @environment_diffs

        nodes = [ @node ]

        roles = [ @role ]

        KnifeSpork::Plugins.run(
          :config => spork_config,
          :hook => hook.to_sym,
          :cookbooks => cookbooks,
          :environments => environments,
          :nodes => nodes,
          :roles => roles,
          :environment_diffs => environment_diffs,
          :environment_path => environment_path,
          :role_path => role_path,
          :node_path => node_path,
          :databag_path => databag_path,
          :cookbook_path => cookbook_path,
          :object_name => @object_name,
          :object_secondary_name => @object_secondary_name,
          :object_difference => @object_difference,
          :ui => ui,
          :args => @args.nil?? {} : @args
        )
      end

      def load_environments_and_cookbook
        ensure_environment_and_cookbook_provided!
       
        if @name_args.size == 2
          environments = @name_args[0].split(",").map{ |env| load_specified_environment_group(env) }
          [ environments.flatten, @name_args[1] ]
        elsif @name_args.size == 1
          [ [default_environments].flatten, @name_args[0] ]
        end
      end

      def verify_and_load_environments
        ensure_environment_provided!

        if @name_args.size == 0
          default_environments
        elsif @name_args.size == 1
          [@name_args[0]]
        end
      end

      def load_specified_environment_group(name)
        if !spork_config.environment_groups.nil? && spork_config.environment_groups.keys.include?(name)
          spork_config.environment_groups[name]
        else
          [name]
        end
      end

      def ensure_environment_and_cookbook_provided!
        if default_environments.empty? && @name_args.size < 2
          ui.error('You must specify an environment or environment group and a cookbook name')
          exit(1)
        end
      end

      def ensure_environment_provided!
        if default_environments.empty? && @name_args.size < 1
          ui.error('You must specify an environment or configure default environments.')
          exit(1)
        end
      end

      def default_environments
        [ spork_config.default_environment || spork_config.default_environments ].flatten.compact
      end

      def pretty_print_json(json)
        options = spork_config[:json_options] || {}
        # generate requires a hash where the keys are symbols
        options = Hash[ options.to_hash.map {|(k,v)| [k.to_sym,v] }] unless options == {}
        JSON.pretty_generate(json, options.to_hash)
      end

      def save_environment_changes(environment, json)
        if spork_config[:environment_path]
          environments_path = spork_config[:environment_path]
        else
          split_cb_path = cookbook_path.split("/")
          environments_path = (split_cb_path[0..-2] << split_cb_path[-1].gsub("cookbooks","environments")).join("/")
        end

        environment_path = File.expand_path( File.join(environments_path, "#{environment}.json") )

        File.open(environment_path, 'w'){ |f| f.puts(json) }
      end

      def save_role_changes(role, json)
        roles_path = File.expand_path( File.join(role_path, "#{role}.json") )

        File.open(roles_path, 'w'){ |f| f.puts(json) }
      end

      def valid_version?(version)
        version_keys = version.split('.')
        return false unless version_keys.size == 3 && version_keys.any?{ |k| begin Float(k); rescue false; else true; end }
        true
      end

      def validate_version!(version)
        if version && !valid_version?(version)
          ui.error("#{version} is not a valid version!")
          exit(1)
        end
      end

      def environment_loader
        @environment_loader ||= Chef::Knife::Core::ObjectLoader.new(::Chef::Environment, ui)
      end

      def role_loader
        @role_loader ||= Chef::Knife::Core::ObjectLoader.new(::Chef::Role, ui)
      end

      # It's not feasible to try and "guess" which cookbook path to use, so we will
      # always just use the first one in the path.
      def cookbook_path
        ensure_cookbook_path!
        [config[:cookbook_path] ||= ::Chef::Config.cookbook_path].flatten[0]
      end

      def environment_path
        if spork_config[:environment_path]
          if spork_config[:environment_path].kind_of?(Array)
            spork_config[:environment_path].first
          else
            spork_config[:environment_path]
          end
        elsif Chef::Config.environment_path && Chef::Config.environment_path.kind_of?(Array)
          Chef::Config.environment_path.first
        elsif Chef::Config.environment_path && Chef::Config.environment_path.kind_of?(String)
          Chef::Config.environment_path
        else
          cookbook_path.gsub("/cookbooks","/environments")
        end
      end

      def role_path
        spork_config[:role_path] || cookbook_path.gsub("/cookbooks","/roles")
      end

      def node_path
        spork_config[:node_path] || cookbook_path.gsub("/cookbooks","/nodes")
      end

      def databag_path
        spork_config[:databag_path] || cookbook_path.gsub("/cookbooks","/data_bags")
      end

      def all_cookbooks
        ::Chef::CookbookLoader.new(::Chef::Config.cookbook_path)
      end

      def load_cookbook(name)
        return name if name.is_a?(Chef::CookbookVersion)

        cookbook = load_from_chef(name) || load_from_berkshelf(name) || load_from_librarian(name)

        cookbook || raise(Chef::Exceptions::CookbookNotFound,
          "Could not find cookbook '#{name}' in any of the sources!")
      end

      def load_from_chef(name)
        all_cookbooks[name]
      rescue Chef::Exceptions::CookbookNotFound,
             Chef::Exceptions::CookbookNotFoundInRepo
        nil
      end

      def load_from_berkshelf(name)
        return unless self.config[:berksfile]
        return unless defined?(::Berkshelf)
        return unless ::File.exist?(self.config[:berksfile])
        berksfile = ::Berkshelf::Berksfile.from_file(self.config[:berksfile])
        lockfile = ::Berkshelf::Lockfile.new(berksfile)

        raise Berkshelf::BerkshelfError, "LockFileNotFound" unless File.exists?(lockfile.filepath)

        cookbook = Berkshelf.ui.mute {
          self.config[:skip_dependencies] ||= false
          berksfile.resolve(lockfile.find(name), {skip_dependencies: self.config[:skip_dependencies]})[:solution].first
        }

        #convert Berkshelf::CachedCookbook to Chef::CookbookVersion
        ::Chef::CookbookLoader.new(File.dirname(cookbook.path))[name]

      end

      # @todo #opensource
      def load_from_librarian(name)
        # Your code here :)
        nil
      end

      def load_cookbooks(cookbook_names)
        cookbook_names = [cookbook_names].flatten
        cookbook_names.collect{ |cookbook_name| load_cookbook(cookbook_name) }
      end


      def load_role_from_file(role_name)
        role_loader.object_from_file("#{role_path}/#{role_name}.json")
      end

      def load_role(role_name)
        Chef::Role.load(role_name)
      end

      def load_node(node)
        Chef::Node.load(node)
      end

      def load_databag(bag)
        Chef::DataBag.load(bag)
      end

      def load_databag_item(bag, item_name)
        Chef::DataBagItem.load(bag, item_name)
      end

      def load_environment(environment_name)
        Chef::Environment.load(environment_name)
      end

      def load_environment_from_file(environment_name)
        begin
          environment_loader.object_from_file("#{environment_path}/#{environment_name}.json")
        rescue FFI_Yajl::ParseError => e
          raise "#{environment_name} environment file has syntactically incorrect json.\n #{e.to_s}" 
        end
      end

      def load_remote_environment(environment_name)
        begin
          Chef::Environment.load(environment_name)
        rescue Net::HTTPServerException => e
          ui.error "Could not load #{environment_name} from Chef Server. You must upload the environment manually the first time."
          exit(1)
        end
      end

      def environment_diff(local_environment, remote_environment)
        local_environment_versions = local_environment.to_hash['cookbook_versions']
        remote_environment_versions = remote_environment.to_hash['cookbook_versions']
        hash_diff remote_environment_versions, local_environment_versions
      end

      def json_diff(a, b)
        pre_json =  JSON.parse(a.respond_to?(:to_json) ? a.to_json : a)
        post_json =  JSON.parse(b.respond_to?(:to_json) ? b.to_json : b)
        Diffy::Diff.new(JSON.pretty_generate(pre_json), JSON.pretty_generate(post_json), :diff=>"-U 3").to_s.gsub(/[()]/, '\\\\\0')
      end

      def hash_diff(hash, other)
        hash.keys.inject({}) do |memo, key|
          unless hash[key] == other[key]
            memo[key] = "#{hash[key]} changed to #{other[key]}"
          end
          memo
        end
      end

      def constraints_diff (environment_diff)
        Hash[Hash[environment_diff.map{|k,v| [k, v.split(" changed to ").map{|x|x.gsub("= ","")}]}].map{|k,v|[k,calc_diff(k,v)]}]
      end

      def calc_diff(cookbook, version)
        components =  version.map{|v|v.split(".")}

        if components.length < 2
          ui.warn "#{cookbook} has no remote version to diff against!"
          return 0
        end

        if components[1][0].to_i != components[0][0].to_i
          return (components[1][0].to_i - components[0][0].to_i)*100
        elsif components[1][1].to_i != components[0][1].to_i
          return (components[1][1].to_i - components[0][1].to_i)*10
        else
          return (components[1][2].to_i - components[0][2].to_i)
        end
      end

      def ensure_cookbook_path!
        if config[:cookbook_path].nil?
          ui.fatal "No default cookbook_path; Specify with -o or fix your knife.rb."
          show_usage
          exit(1)
        end
      end

      def unload_berkshelf_if_specified
        # Temporary fix for #138 to allow Berkshelf functionality
        # to be bypassed until #85 has been completed and Berkshelf 3 support added
        if spork_config.skip_berkshelf
          ui.warn "Unloading Berkshelf as skip_berkshelf option found in config"
          Object.send(:remove_const, :Berkshelf) if defined?(::Berkshelf)
        end
      end
    end

    def self.included(receiver)
      receiver.extend(ClassMethods)
      receiver.send(:include, InstanceMethods)
    end

    private
    def load_paths
      [ File.expand_path("#{cookbook_path.gsub('cookbooks','')}/config/spork-config.yml"), File.expand_path('config/spork-config.yml'), '/etc/spork-config.yml', File.expand_path('~/.chef/spork-config.yml'), File.expand_path("~/spork-config.yml") ]
    end
  end
end
